tools/c7n_azure/c7n_azure/resources/cosmos_db.py

```diff
@@ -52,6 +74,193 @@ class resource_type(ArmResourceManager.resource_type):
+        data_client = CosmosDB.get_data_client(parent_resource, self)
+
+        try:
+            databases = list(data_client.ReadDatabases())
+        except HTTPFailure as e:
+            if e.status_code == 403:
+                log.error("Identity does not have permissions to read CosmosDB accounts"
+                          " or firewall is blocking access.")
+            else:
+                print(e)
+            return None
+
+        collections = []
+
+        for d in databases:
+            container_result = list(data_client.ReadContainers(d['_self']))
+            for c in container_result:
+                c.update({'c7n:document-endpoint':
+                         parent_resource.get('properties').get('documentEndpoint')})
+                collections.append(c)
+
+        return collections
+
+
+@CosmosDBCollection.filter_registry.register('offer')
+@CosmosDBDatabase.filter_registry.register('offer')
+class CosmosDBOfferFilter(ValueFilter):
+    schema = type_schema('offer', rinherit=ValueFilter.schema)
+    schema_alias = True
+
+    def process(self, resources, event=None):
+        futures = []
+        results = []
+
+        # Group all resources by account because offers are queried per account not per collection
+        account_sorted = sorted(resources, key=CosmosDBOfferFilter.account_key)
+        account_grouped = [list(it) for k, it in groupby(
+            account_sorted,
+            CosmosDBOfferFilter.account_key)]
+
+        # Process database groups in parallel
+        with self.executor_factory(max_workers=3) as w:
+            for resource_set in account_grouped:
+                futures.append(w.submit(self.process_resource_set, resource_set))
+
+            for f in as_completed(futures):
+                if f.exception():
+                    self.log.warning(
+                        "Diagnostic settings filter error: %s" % f.exception())
+                    continue
+                else:
+                    results.extend(f.result())
+
+            return results
+
+    def process_resource_set(self, resources):
+        matched = []
+
+        try:
+            # Skip if offer key is present anywhere because we already
+            # queried and joined offers in a previous filter instance
+            if not resources[0].get('c7n:offer'):
+
+                # Get the data client keys
+                parent_key = resources[0]['c7n:parent-id']
+                key = CosmosDB.get_cosmos_key(
+                    ResourceIdParser.get_resource_group(parent_key),
+                    ResourceIdParser.get_resource_name(parent_key),
+                    self.manager)
+
+                # Build a data client
+                data_client = CosmosClient(
+                    url_connection=resources[0]['c7n:document-endpoint'],
+                    auth={'masterKey': key})
+
+                # Get the offers
+                offers = list(data_client.ReadOffers())
```

Comment:
Is it the same list of offers for all CosmosDBs in the subscription?

Reply:
The offers are per CosmosDB account - so in this threading implementation `process_resource_set` only ever receives groups of collections or databases in the same account.  This gives us parallelism without overly complicating dealing with the offers.